| Dataset Name   | Year | Dataset Acc.    | BioProject                                                         | Paper URL                                                                                       | No. of Samples | Number of Cells | Number of cell types |
| -------------- | ---- | --------------- | ------------------------------------------------------------------ | ----------------------------------------------------------------------------------------------- | -------------- | --------------- | -------------------- |
| Fan10X         | 2019 | GSE118127       | [PRJNA484542](https://www.ncbi.nlm.nih.gov/bioproject/PRJNA484542) | [nature](https://www.nature.com/articles/s41467-019-11036-9)                                    | 31             | 20,676          | 19                   |
| Wagner10X      | 2020 | E-MTAB-8381     | PRJEB34869                                                         | [nature](https://www.nature.com/articles/s41467-020-14936-3)                                    | 4              | 12,160          | 6                    |
| Lengyel10X     | 2022 | EGAS00001006780 | \-                                                                 | [sciencedirect](https://www.sciencedirect.com/science/article/pii/S2211124722017302?via%3Dihub) | 4              | 22,332          | 6                    |
| LengyelDropSeq | 2022 | \-              | \-                                                                 | [sciencedirect](https://www.sciencedirect.com/science/article/pii/S2211124722017302?via%3Dihub) | 2              | 3,802           | 6                    |
# Group 2
These are datasets where paired fastq reads are not readily availible, but other formats are that we can use.

- Fan10X dataset has fastq files availible, but they were not split into R1 and R2, so could not go through the cellranger pipeline. We can use the BAM files instead and either use BAM2FASTQ then cellranger, or just quantify the BAM files. We can still use the SRA table as metadata.
- Wagner10X has fastq files availible from [EBI](https://www.ebi.ac.uk/biostudies/arrayexpress/studies/E-MTAB-8381) rather than SRA. These can be processed in a similar manner to the group 1 datasets, but the metadata will need curating manually.
- For Lengyel10X and LengyelDropSeq, the raw fastq data is still not availible, but keep checking [EGA](https://ega-archive.org/datasets/EGAD00001010076) in case they add it. Instead, we can download the processed RDS or H5AD file from [cellxgene](https://cellxgene.cziscience.com/collections/d36ca85c-3e8b-444c-ba3e-a645040c6185) - this contains both the 10X and dropseq data lumped together. We can load this into a Seurats object and then filter to separate them into 2 datasets. Metadata will need adjusting manually. Use [cellxgene explorer](https://cellxgene.cziscience.com/e/d1207c81-7309-43a7-a5a0-f4283670b62b.cxg/) to visualise which samples come from which donors and techniques, etc.
